crypto/src/mls/credential/mod.rs

```diff
@@ -228,17 +230,37 @@ mod tests {
         let alice_identifier = ClientIdentifier::X509(HashMap::from([(case.signature_scheme(), certs)]));
 
         let (bob_identifier, _) = x509_test_chain.issue_simple_certificate_bundle("bob", None);
-        assert!(matches!(
-            try_talk(&case, Some(&x509_test_chain), alice_identifier, bob_identifier)
-                .await
-                .unwrap_err(),
-            CryptoError::InvalidIdentity
-        ));
+        let err = try_talk(&case, Some(&x509_test_chain), alice_identifier, bob_identifier)
+            .await
+            .unwrap_err();
+        // assert!(innermost_source_matches!(err, Error::InvalidIdentity));
```

Comment:
This is still a mystery?

Reply:
Not a mystery anymore, but I'm still figuring out how to solve it. In this particular case, we have `dbg!` output that looks like this:

```
[crypto/src/mls/credential/mod.rs:236:9] &err = Recursive(
    Mls {
        context: "initializing mls",
        source: Recursive(
            MlsClient {
                context: "initializing mls client",
                source: Recursive(
                    MlsCredential {
                        context: "getting client id",
                        source: InvalidIdentity,
                    },
                ),
            },
        ),
    },
)
```

What's going on here is that `RecursiveError` boxes up its sources, so that the size of the enum doesn't go infinite. So even though the representation of the innermost source is just `InvalidIdentity`, what's actually sitting there is a `Box<core_crypto::mls::credential::Error>` with the `InvalidIdentity` variant. 

[`<impl<T> Error for Box<T> where T: Error>::source`](https://doc.rust-lang.org/stable/src/alloc/boxed.rs.html#3028-3030) just defers to the inner `.source()` implementation. So in the case of this error, it returns `None`, because `InvalidIdentity` has no source error. 

The `innermost_source_matches!` macro uses that passed type--in this case, `core_crypto:mls::credential::Error`--to determine what `downcast_ref` should be looking for. And `downcast_ref` decides, correctly, that `core_crypto:mls::credential::Error` and `Box<core_crypto:mls::credential::Error>` are of different types, and refuses to downcast one as the other, which means of course that the match fails.

Reply:
[Deref patterns](https://github.com/rust-lang/rust/issues/87121) would be really useful here.

Reply:
fd40ffc6